File: android/src/main/kotlin/org/thelittlehackers/network_settings_state/NetworkSettingsStatePlugin.kt
package org.thelittlehackers.network_settings_state

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.net.wifi.WifiManager
import android.telephony.TelephonyManager
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/**
 * Reports whether the Wi-Fi and mobile data radios are enabled in the
 * device's system settings, independent of which network is currently
 * carrying traffic.
 */
class NetworkSettingsStatePlugin :
    FlutterPlugin,
    MethodCallHandler,
    EventChannel.StreamHandler {

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private lateinit var applicationContext: Context

    private var wifiStateReceiver: BroadcastReceiver? = null
    private var eventSink: EventChannel.EventSink? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = binding.applicationContext

        methodChannel = MethodChannel(binding.binaryMessenger, "network_settings_state")
        methodChannel.setMethodCallHandler(this)

        eventChannel = EventChannel(binding.binaryMessenger, "network_settings_state_stream")
        eventChannel.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        unregisterWifiStateReceiver()
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "isWifiEnabled" -> result.success(isWifiEnabled())
            "isMobileDataEnabled" -> result.success(isMobileDataEnabled())
            "getSnapshot" -> result.success(currentSnapshot())
            else -> result.notImplemented()
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
        registerWifiStateReceiver()
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
        unregisterWifiStateReceiver()
    }

    private fun isWifiEnabled(): Boolean {
        val wifiManager = applicationContext
            .getSystemService(Context.WIFI_SERVICE) as? WifiManager
            ?: return true

        return wifiManager.isWifiEnabled
    }

    /**
     * Returns whether mobile data is enabled in device settings, or
     * `null` if this cannot be determined (no SecurityException
     * fallback beyond fail-open `true`, chosen deliberately: a missing
     * permission should not present as "mobile data disabled" to the
     * end user, since that is a distinct and misleading failure mode).
     */
    private fun isMobileDataEnabled(): Boolean {
        val telephonyManager = applicationContext
            .getSystemService(Context.TELEPHONY_SERVICE) as? TelephonyManager
            ?: return true

        return try {
            telephonyManager.isDataEnabled
        } catch (securityException: SecurityException) {
            // READ_PHONE_STATE not granted on API < 29. Fail open rather
            // than reporting a false "disabled" state.
            true
        }
    }

    private fun currentSnapshot(): Map<String, Any?> {
        return mapOf(
            "wifiEnabled" to isWifiEnabled(),
            "mobileDataEnabled" to isMobileDataEnabled(),
        )
    }

    private fun registerWifiStateReceiver() {
        if (wifiStateReceiver != null) {
            return
        }

        val receiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                eventSink?.success(currentSnapshot())
            }
        }

        applicationContext.registerReceiver(
            receiver,
            IntentFilter(WifiManager.WIFI_STATE_CHANGED_ACTION),
        )

        wifiStateReceiver = receiver
    }

    private fun unregisterWifiStateReceiver() {
        wifiStateReceiver?.let {
            applicationContext.unregisterReceiver(it)
        }
        wifiStateReceiver = null
    }
}

File: ios/network_settings_state/Sources/network_settings_state/NetworkSettingsStatePlugin.swift
import Flutter
import Network
import UIKit

/// Reports whether the Wi-Fi radio appears enabled on iOS, and always
/// reports `nil` for mobile data since no public API exists to query
/// the global device setting.
///
/// The Wi-Fi reading is best-effort: it reflects whether a Wi-Fi
/// interface is currently available via `NWPathMonitor`, which is a
/// close but imperfect proxy for the Settings toggle. See the package
/// README for details.
public class NetworkSettingsStatePlugin: NSObject, FlutterPlugin, FlutterStreamHandler {

    private let pathMonitor = NWPathMonitor()
    private let monitorQueue = DispatchQueue(label: "org.thelittlehackers.network_settings_state.monitor")
    private var eventSink: FlutterEventSink?

    override init() {
        super.init()

        // Started unconditionally and kept running for the plugin's
        // entire lifetime. NWPathMonitor.currentPath only reflects
        // real state once start() has been called at least once --
        // previously this only happened when a stream listener
        // subscribed via onListen, so a one-shot isWifiEnabled() /
        // getSnapshot() call made before any stream subscription would
        // read an unstarted, unsatisfied path and incorrectly report
        // Wi-Fi as disabled even when it was enabled and connected.
        pathMonitor.start(queue: monitorQueue)
    }

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = NetworkSettingsStatePlugin()

        let methodChannel = FlutterMethodChannel(
            name: "network_settings_state",
            binaryMessenger: registrar.messenger()
        )
        registrar.addMethodCallDelegate(instance, channel: methodChannel)

        let eventChannel = FlutterEventChannel(
            name: "network_settings_state_stream",
            binaryMessenger: registrar.messenger()
        )
        eventChannel.setStreamHandler(instance)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "isWifiEnabled":
            result(isWifiEnabled())
        case "isMobileDataEnabled":
            // No public API exists on iOS to query the global mobile
            // data toggle. Returning nil rather than guessing.
            result(nil)
        case "getSnapshot":
            result(currentSnapshot())
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func isWifiEnabled() -> Bool {
        let path = pathMonitor.currentPath
        return path.availableInterfaces.contains { $0.type == .wifi }
    }

    private func currentSnapshot() -> [String: Any?] {
        return [
            "wifiEnabled": isWifiEnabled(),
            "mobileDataEnabled": nil,
        ]
    }

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events

        // The monitor is already running (started in init); just hook
        // up the update handler to forward changes to the stream.
        pathMonitor.pathUpdateHandler = { [weak self] _ in
            guard let self = self else { return }
            DispatchQueue.main.async {
                self.eventSink?(self.currentSnapshot())
            }
        }

        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        // Deliberately does NOT call pathMonitor.cancel() -- the
        // monitor must keep running for the plugin's lifetime so that
        // one-shot method channel calls (isWifiEnabled, getSnapshot)
        // continue to work after a stream listener unsubscribes.
        pathMonitor.pathUpdateHandler = nil
        eventSink = nil
        return nil
    }
}
